### PR TITLE
Feature cambiar foto de grupo 160675744

### PR DESCRIPTION
**Pull request by jpducassou**
Added image upload support for group creation and update.
Extracted most of the functionality from user profiles to a Rails concern mixin.

### CODE REVIEW

File: app/models/group.rb [hunk 1]
```diff
@@ -5,6 +5,10 @@ class Group < ApplicationRecord
   has_many :user_groups, dependent: :destroy
   has_many :users, through: :user_groups
 
+  validates :image_url, length: { maximum: 250 }, allow_nil: true, allow_blank: true, format: \
+  { with: %r{https?:\/\/[\S]+} }
+  has_one_attached :image
+
```
Comment (joaquinvillar) [inline app/models/group.rb]: nit: unnecessary line

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -98,40 +93,23 @@ def user_params_update
...
 
-    def purge_stale_picture(user_params)
-      return if user_params['image'].nil?
-      return if user_params['image'] == ""
-
-      @user.profile_picture.purge
+    def purge_image
+      purge_stale_image(@user.profile_picture, @user_params, 'image')
     end
+
```
Comment (joaquinvillar) [inline app/controllers/users_controller.rb]: nit: unnecessary line

File: app/controllers/concerns/image_upload.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
+require 'mime/types'
+require 'base64'
+
+module ImageUpload
+  extend ActiveSupport::Concern
+
+  def validate_image_field(request_params, image_field)
+    return if request_params[image_field].nil?
+    return if %r{^https?:\/\/}.match?(request_params[image_field])
```
Comment (joaquinvillar) [inline app/controllers/concerns/image_upload.rb]: is this needed if you have the validation in the model?

File: app/controllers/concerns/image_upload.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    raise StandardError, 'Invalid image field!'
+  end
+
+  def update_image_field(image_field, image_info)
+    return unless image_info
+    return unless image_info[:mime]
+    return unless image_info[:data]
+
+    extension = MIME::Types[image_info[:mime]].first.preferred_extension
+    image_file = StringIO.new(Base64.decode64(image_info[:data]))
```
Comment (joaquinvillar) [inline app/controllers/concerns/image_upload.rb]: what happens if the data is not valid?

File: app/controllers/concerns/image_upload.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    image_field.attach(io: image_file, filename: "avatar.#{extension}")
+    true
+  end
+
+  def purge_stale_image(image_field, request_params, image_field_name)
+    return if request_params[image_field_name].nil?
+    return if request_params[image_field_name] == ""
```
Comment (joaquinvillar) [inline app/controllers/concerns/image_upload.rb]: you can do in one line with blank?  leave you the references
https://apidock.com/rails/Object/blank%3F

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -1,10 +1,13 @@
...
 class UsersController < ApplicationController
+  include ImageUpload
+
   before_action :set_user_from_token, only: [:hide_alert]
   before_action :set_user_from_route, only: [:show, :update]
 
+  before_action :set_image_info, only: [:update]
+  after_action  :update_image, only: [:update]
+  after_action  :purge_image, only: [:update]
+
```
Comment (joaquinvillar) [inline app/controllers/users_controller.rb]: nit: unnecessary line

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -28,27 +31,19 @@ def show
...
     # Validate user id
     check_id_match(@user.id)
 
-    # Validate user params
-    user_params = user_params_update
-
-    # check if image is url or data
-    validate_image_field(user_params)
-
     User.transaction do
```
Comment (joaquinvillar) [inline app/controllers/users_controller.rb]: Is this transaction needed if you only have one action?

File: app/controllers/groups_controller.rb [hunk 8]
```diff
@@ -26,17 +30,18 @@ def create
...
     @current_users = @group.users
     @admin = @group.admin_id
-    @new_params = group_params
     destroy_associations(@current_users, @admin, @new_params)
 
-    if @group.update(@new_params)
-      render json: @group.as_json(include: :user_groups)
-    else
-      render json: @group.errors, status: :unprocessable_entity
+    Group.transaction do
```
[the remark this answers is not in the document]
Comment (jpducassou) [reply app/controllers/groups_controller.rb]: Fixed!

**Pull request merged**